conda_libmamba_solver/exceptions.py

```diff
@@ -15,6 +22,44 @@ def __init__(self, message, **kwargs):
         super(UnsatisfiableError, self).__init__(str(message))
 
 
+class RequestedAndPinnedError(SpecsConfigurationConflictError):
+    """
+    Raised when a spec is both requested and pinned.
+    """
+
+    def __init__(self, requested_specs, pinned_specs, prefix):
+        message = (
+            dedent(
+                """
+                Requested specs overlap with pinned specs.
+                  requested specs: {requested_specs_formatted}
+                  pinned specs: {pinned_specs_formatted}
+
+                Consider adjusting your requested specs to respect the pin(s),
+                or explicitly remove the offending pin(s) from the configuration.
+                Use 'conda config --show-sources' to look for 'pinned_specs'.
+                Pinned specs may also be defined in the file
+                {pinned_specs_path}.
```

Comment:
rerunning my test from [here](https://github.com/conda/conda-libmamba-solver/pull/289#issuecomment-1729832553) I ran into this error, but in my opinion it doesn't make sense.

```
#15 [11/12] RUN conda install -c conda-forge -n myenv --solver libmamba --file requirements.txt
#15 1.161 
#15 1.161 RequestedAndPinnedError: Requested specs overlap with pinned specs.
#15 1.161   requested specs: 
#15 1.161     - pandas==1.2.5=py37h295c915_0
#15 1.161     - python==3.7.3
#15 1.161     - scipy==1.7.3=py37hf2a6cf1_0
#15 1.161   pinned specs: 
#15 1.161     - pandas==1.2.5
#15 1.161     - python==3.7.3
#15 1.161     - scipy==1.7.3
#15 1.161 
#15 1.161 Consider adjusting your requested specs to respect the pin(s),
#15 1.161 or explicitly remove the offending pin(s) from the configuration.
#15 1.161 Use 'conda config --show-sources' to look for 'pinned_specs'.
#15 1.161 Pinned specs may also be defined in the file
#15 1.161 /opt/conda/envs/myenv/conda-meta/pinned.
```

in this case I'm not sure it's fair to say there's overlap when the requested and pinned specs are perfectly compatible.


> Consider adjusting your requested specs to respect the pin(s)

This suggestion does not apply - my requests and the pins respect each other :)

Reply:
Overlap here means that package names already covered by the pins are present (in one way or another) in the CLI specs. Testing for "spec compatibility" is tricky to generalize robustly as explained in the PR description, so in this PR I've opted for an approach where we prevent it altogether. I'm assuming that if someone wants to specify a spec in the CLI that is already pinned, the only legitimate use case is to further restrain it. In those cases, I'd say the user should modify the pinned file to apply those restraints there. 

That's my assumption, but I am happy to hear other points of view and reconsider my position.

Reply:
@wuillaum - can you check https://github.com/conda/conda-libmamba-solver/pull/292? Maybe that's more in line with what you expect.

Reply:
@jaimergp the branch for that PR seems to work as I would expect.

Success as expected (and in half the time as the classic solver!):
```
#14 [10/12] RUN conda install -c conda-forge -n myenv --solver classic  --file requirements.txt
#14 0.911 Collecting package metadata (current_repodata.json): ...working... done
#14 28.08 Solving environment: ...working... done
#14 29.34 
#14 29.34 # All requested packages already installed.
#14 29.34 
#14 DONE 30.6s

#15 [11/12] RUN conda install -c conda-forge -n myenv --solver libmamba --file requirements.txt
#15 1.411 Channels:
#15 1.411  - conda-forge
#15 1.411  - defaults
#15 1.411 Platform: linux-64
#15 1.411 Collecting package metadata (repodata.json): ...working... done
#15 13.62 Solving environment: ...working... done
#15 15.09 
#15 15.09 # All requested packages already installed.
#15 15.09 
#15 DONE 15.2s
```


Failures as expected:
```
 > [12/12] RUN conda install -n myenv --solver libmamba pandas=1.3.5:
13.74     - pandas=1.3.5
13.74   pinned specs: 
13.74     - pandas==1.2.5
13.74     - python==3.7.3
13.74     - scipy==1.7.3
13.74 Use 'conda config --show-sources' to look for 'pinned_specs' and 'track_features'
13.74 configuration parameters.  Pinned specs may also be defined in the file
13.74 /opt/conda/envs/myenv/conda-meta/pinned.
```

Reply:
294 is now part of this PR.

Reply:
Oops, sorry the thread was not up-to-date. Ok, awesome, thanks for confirming. We also have tests for this behaviour now (and profiles) but seeing it in other machines is always refreshing. Resolving this conversation. Thanks again for the valuable feedback!